src/collector.rs

```diff
@@ -0,0 +1,9 @@
+// TODO: Document
+
+use std::borrow::Cow;
+
+use crate::registry::Descriptor;
+
+pub trait Collector<M: Clone> {
+    fn collect<'a>(&'a self) -> Box<dyn Iterator<Item = (Cow<Descriptor>, Cow<M>)> + 'a>;
+}
```

Comment:
This is the trait one would need to implement to provide a custom collector, e.g. a process collector.

Reply:
I see this trait is generic over `M`, which I think is the type of the metric? How does the `Clone` bound work with collectors/registries of type `Box<dyn EncodeMetric>` which aren't `Clone`? 🤔

Reply:
> How does the `Clone` bound work with collectors/registries of type `Box<dyn EncodeMetric>` which aren't `Clone`? thinking

It doesn't. For now this is just a proposal. We could require `Clone` on `EnocdeMetric`, though I think the much better idea is to no longer be generic over the metric type. See https://github.com/prometheus/client_rust/pull/82#discussion_r956892586.

Let me know what you think @sd2k!

Reply:
Ah great, that makes sense. I agree that switching to an enum instead of generics/trait objects would simplify things a lot! (Especially since adding a `Clone` bound to `EncodeMetric` would mean it wasn't object safe, so I don't think we could use trait objects anyway).

Reply:
> (Especially since adding a Clone bound to EncodeMetric would mean it wasn't object safe, so I don't think we could use trait objects anyway).

Good point. We would have to drop the `Cow<M>` in favor of `M`, but then again, let's try to get rid of `M`.